### PR TITLE
Add package: gn

### PR DESCRIPTION
**Pull request by thunder-coding**
TODOs:
- [x] Add gn
- [x] Test if it works
   Sample project compile well on aarch64 Android 11
- [x] Create a `termux_setup_gn` script
~~- [ ] Add a package that requires gn to build to test `termux_setup_gn`~~
~~- [ ] Add package `v8`~~
~~- [ ] `nodejs` link to shared `v8`~~

### CODE REVIEW

**Comment by SPECIFIC10** [comment]
I need some tool to hack phone number

**Comment by thunder-coding** [comment]
Its picking up the correct linker
```
cxx = aarch64-linux-android-clang++
ar = aarch64-linux-android-ar
ld = aarch64-linux-android-clang++
```
Notice `-clang++` at the end instead of `-ld`

**Comment by SPECIFIC10** [comment]
I need some tool for number hacking plz

**Comment by thunder-coding** [comment]
The compiled `gn` works well and for testing `termux_setup_gn` script, I will be adding the v8 package and make nodejs link to that v8. The v8 version will be kept roughly near to nodejs' in order to reduce number of patches.

Also while building Node.js 16, v8 seems  to compile for wrong arch, so having shared v8 makes sense. See suhan-paradkar/termux-packages#8

File: scripts/build/setup/termux_setup_gn.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+termux_setup_gn() {
+	termux_setup_ninja
+	local GN_COMMIT=eea3906f0e2a8d3622080127d2005ff214d51383
+	local GN_FOLDER=$TERMUX_COMMON_CACHEDIR/gn-$GN_COMMIT
+	local GN_TARFILE=$TERMUX_COMMON_CACHEDIR/gn_$GN_COMMIT.tar.gz
+	local GN_SOURCE=https://gn.googlesource.com/gn/+archive/$GN_COMMIT.tar.gz
+	test ! -f $GN_TARFILE && termux_download \
```
Comment (xeffyr) [inline scripts/build/setup/termux_setup_gn.sh:7]: `test ! -f $GN_TARFILE` is not necessary there.

`termux_download` will re-download it only if integrity check fails or file does not exist. Here you are disabling integrity checking of existing file.

File: scripts/build/setup/termux_setup_gn.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		$GN_SOURCE \
+		$GN_TARFILE \
+		SKIP_CHECKSUM
+	mkdir -p $GN_FOLDER
+	tar xf $GN_TARFILE -C $GN_FOLDER
```
Comment (xeffyr) [inline scripts/build/setup/termux_setup_gn.sh:12]: You are extracting tar without checking the presence of already extracted installation.

**Review by xeffyr — CHANGES_REQUESTED** [review]
Please make sure you have checked how `termux_setup_golang` or `termux_setup_cmake` work and adjust `termux_setup_gn` accordingly.

Specifically:
* Need integrity check of downloaded tarball. No matter whether it has been downloaded and checked previously. You need to verify it always.
* Do not re-install gn if it was already installed.

**Review by xeffyr — APPROVED** [review]

**Comment by xeffyr** [comment]
Fails at
```
ERROR at //BUILD.gn:5285:3: Assertion failed.
  assert(!is_component_build)
  ^-----
```

Possibly you need to add `is_component_build=false`.

**Comment by suhan-paradkar** [comment]
It requires libicu now??

**Comment by thunder-coding** [comment]
> It requires libicu now?? 

It requires icu for i18n support

**Comment by hnmn** [comment]
ig `target_os="android"` be correct 
Also `target_os = ['android'] ` need in .gclient

**Comment by thunder-coding** [comment]
I got it why v8 fails with shared v8, libicu is missing pkg configs, I will push the fix soon.

And I got v8 building for x86_64, aarch64 build is going on.

If it passes, I will add patches for v8 which will be stealed from nodejs vi patches

**Comment by suhan-paradkar** [comment]
Well.. 32 bit CPUs are facing few strange problems with gn..

**Comment by thunder-coding** [comment]
> Well.. 32 bit CPUs are facing few strange problems with gn..

Probably that's due to some variables defined in `scripts/build/termux_step_setup_toolchain.sh`

Will try unsetting all if them after CI for aarch64 and x86_64 pass

**Comment by suhan-paradkar** [comment]
Well... About aarch64 build... Looks like the CI ran out of space...

**Comment by suhan-paradkar** [comment]
Same about x86_64 too

**Comment by thunder-coding** [comment]
Now, that I have revised libicu to include pkg-config, I can confirm that V8 is looking for icu-i18n.pc at a really wrong place

https://github.com/chromium/chromium/blob/be8ee8535827682b07c0864c165eb9bca4005521/build/config/sysroot.gni#L20

**Comment by thunder-coding** [comment]
https://github.com/chromium/chromium/blob/master/docs/clang.md

Seems like V8 uses a different clang binary than what we expect to compile stuff with.

**Comment by xeffyr** [comment]
@thunder-coding The problem is `--sysroot=../../../../../../../data/data/com.termux/files`. Termux prefix is not valid compiler system root. There should be something like `--sysroot=$TERMUX_STANDALONE_TOOLCHAIN/sysroot` instead. Any headers and library paths of dependencies should be included through CFLAGS and LDFLAGS respectively.

**Comment by thunder-coding** [comment]
The input errors seem to be prominent even after setting sysroot to `$TERMUX_STANDALONE_TOOLCHAIN/sysroot`

**Comment by hnmn** [comment]
`v8_enable_i18n_support=false`  try this

**Comment by thunder-coding** [comment]
> `v8_enable_i18n_support=false`  try this 

This works, but at the cost of i18n support which is not desired.

**Comment by hnmn** [comment]
Can we just follow the official build guide for cross compiling v8?
Ig gn is looking for ndk at $v8_root/third_party

**Comment by thunder-coding** [comment]
V8 seems to have a lot of problems compiling for Termux. Should I skip adding it? Atleast we can have GN for those who would like to have it...

**Comment by thunder-coding** [comment]
Okay, so I have removed all my commits for V8. I might try if I can get it building in another PR. And about the `termux_setup_gn` script, I am sure it will work, but for evey package that requires GN to build, we may need to patch respective `BUILD.gn` file of the same in order to make it choose correct build tools because GN doesn't by itself choose proper build tools from environment variables.

**Review by xeffyr — APPROVED** [review]

**Pull request merged**